rpc: fix getrawtransaction segfault

mzumsande opened the pull request:
The crash, reported in #28986, happens when calling `getrawtransaction` for any mempool transaction with `verbosity=2`, while pruning, because the rpc calls `IsBlockPruned(const CBlockIndex* pblockindex)`, which dereferences `pblockindex` without a check.

For ease of backporting this PR fixes it just locally in `rpc/rawtransaction.cpp` by moving the check for`!blockindex` up so that `IsBlockPruned()` will not be called with a `nullptr`. We might also want to change `IsBlockPruned()` so it doesn't crash when called with a `nullptr`, but I didn't do that here.

Fixes #28986

mzumsande committed:
rpc: fix getrawtransaction segfault

The crash would happen when querying a mempool transaction with verbosity=2, while pruning.

nogamike76 commented:
No it wouldn’t. What needs to be done

May God guide us and may we all, including myself find happiness, love and
forgiveness. Be kind to someone, help someone with something that would
help them. Make sure you tel your family you love them. May the kings be
protected and may our King, Jesus Christ, always be the divine and glory,
in honor of our father, the almighty one, God.
Spirit Leader
Michael Noga jr.
Descendant of King David
Nogah, son of David
T8<3
$€£¥


On Tue, Dec 5, 2023 at 11:53 AM maflcko ***@***.***> wrote:

> lgtm ACK 494a926
> <https://github.com/bitcoin/bitcoin/commit/494a926d05df44b60b3bc1145ad2a64acf96f61b>
>
> A test wouldn't hurt, I'd say.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/29003#issuecomment-1841418135>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AQXNDFJ6XO7O4OQRB2MLQ5TYH5USLAVCNFSM6AAAAABAIENTZCVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQNBRGQYTQMJTGU>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

pablomartin4btc reviewed (approved):
tACK 494a926d05df44b60b3bc1145ad2a64acf96f61b

I've managed to reproduce the issue, this PR fixes it.

mzumsande commented:
> A test wouldn't hurt, I'd say.

Will add one soon.

pablomartin4btc commented:
> > We might also want to change `IsBlockPruned()` so it doesn't crash when called with a `nullptr`, but I didn't do that here.
> 
> I think we should avoid the use of pointers when they are assumed to not be null. In C++, references exit. However, this can be done in a follow-up.

Sorry, why not fixing the issue directly here(?):
```diff
--- a/src/node/blockstorage.cpp
+++ b/src/node/blockstorage.cpp
@@ -585,7 +585,7 @@ const CBlockIndex* BlockManager::GetLastCheckpoint(const CCheckpointData& data)
 bool BlockManager::IsBlockPruned(const CBlockIndex* pblockindex)
 {
     AssertLockHeld(::cs_main);
-    return (m_have_pruned && !(pblockindex->nStatus & BLOCK_HAVE_DATA) && pblockindex->nTx > 0);
+    return (m_have_pruned && (pblockindex != nullptr) && !(pblockindex->nStatus & BLOCK_HAVE_DATA) && pblockindex->nTx > 0);
 }
```
Or even `return false;` earlier before the assert.

mzumsande commented:
> Sorry, why not fixing the issue directly here(?):

I initially decided against this because
1) calling `IsBlockPruned()` doesn't make any sense in the first place when there is no block.
2) There were lots of refactors in the blockstorage module over the last year and I wasn't sure how cleanly this would backport.

I think that it's not one or the other, both changes make sense. However, to avoid the question if `IsBlockPruned()` should return true or false for a `nullptr`, it would probably be better change it to use a reference as [maflcko](https://github.com/maflcko) suggested.

achow101 commented:
Perhaps a test for this case?

mzumsande committed:
test: add regression test for the getrawtransaction segfault

This fails on master without the previous commit.

mzumsande commented:
I added a regression test (that will segfault on master). In order to trigger this, it's not enough to enable pruning, you have to actually have pruned a block, so I did this with `-fastprune` in the test. Because this will probably seem a bit random to someone reading the test without this context, I added a comment linking to this PR.

theStack commented:
Concept ACK

theStack reviewed (approved):
Tested ACK 9075a446461ccbc446d21af778aac50b604f39b3

Reproduced the crash on master with my pruned node manually via
```
$ ./src/bitcoin-cli getrawmempool | head -n 2
[
    "sometxid...",
$ ./src/bitcoin-cli getrawtransaction sometxid... 2
error: timeout on transient error: Could not connect to the server 127.0.0.1:8332 (error code 1 - "EOF reached")

Make sure the bitcoind server is running and that you are connecting to the correct RPC port.

-----
bitcoind crashes with "Segmentation fault (core dumped)"
```
and verified that the PR fixes it. Also checked that the test crashes if cherry-picked on master, and succeeds in the PR branch.

Brock124590 commented on src/rpc/rawtransaction.cpp:
```diff
@@ -396,7 +396,7 @@ static RPCHelpMan getrawtransaction()
     }
-    if (verbosity == 1) {
+    if (verbosity == 1 || !blockindex) {
         TxToJSON(*tx, hash_block, result, chainman.ActiveChainstate());
```
Do you have your secret phrase

fanquake commented:
Added to #29011 for backporting to 26.x.

pablomartin4btc reviewed (commented):
ACK 9075a446461ccbc446d21af778aac50b604f39b3 on new reg test [added](https://github.com/bitcoin/bitcoin/pull/29003#issuecomment-1841662429)

fanquake commented:
Added to #28768 for backporting to 25.x.
